### PR TITLE
Add font-end for comments

### PR DESCRIPTION
**Pull request by pohlinwei**

### CODE REVIEW

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+class Comment {
+  /**
+   * @param {string} commenterName 
+   * @param {string} commentMsg Commenter's message.
+   * @param {number} sentiment Either 1 (positive) or -1 (negative).
+   * @param {RepliesManager} repliesManager Manager that manages all associated replies.
+   */
+  constructor(commenterName, commentMsg, sentiment, repliesManager) {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/comment.js:8]: Shall we create a enum for sentiment instead? I can see that right now we have only two possible result of sentiment, but in case we want to have more sentiments, using a number doesn't seem flexible enough?

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  static toComment(commentJson) {
+    const commenterName = commentJson.name;
+    const commentMsg = commentJson.commentMsg;
+    const commentSentiments = commentJson.sentiment;
+    const repliesManager = RepliesManager.toRepliesManager(commentJson.repliesManager);
+    return new Comment(commenterName, commentMsg, commentSentiments, repliesManager);
+  }
+
+  /** @return {HTML element} HTML element representing this comment */
+  get toElement() {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/comment.js:25]: Possibly a naming issue. This method doesn't seem to be a getter only? It is mutating html elements on page.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    commentElement.appendChild(replyForm);
+    commentElement.appendChild(replies);
+
+    this.enableToggleReplyForm(commentElement, replyForm);
+    this.enableToggleReplies(commentElement, replies);
+    return commentElement;
+  }
+
+  /** @return {HTML as string} HTML for commenter's name, comment message and view replies button. */
+  get mainSectionHtml() {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/comment.js:41]: Where is this getter used? I don't see it being used anywhere

File: portfolio/src/main/webapp/projects/commentart/scripts/utilities.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+function enableAsyncForm(form, getInputs, callBackFunc) {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/utilities.js]: Missing doc on public methods, especially util methods.

callBackFunc isn't a good name, you can use something like `onDataFetched`

File: portfolio/src/main/webapp/projects/commentart/scripts/replyform.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    formElement.setAttribute('name', 'reply-form');
+    const SERVER_URL = '/add_reply';
+    formElement.setAttribute('action', SERVER_URL);
+    formElement.setAttribute('method', 'post');
+
+    formElement.innerHTML = this.getFormContent();
+    this.enableAsyncReply(formElement, addReply);
+  }
+  /** Gets HTML for reply form. */
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/replyform.js:20]: nit: missing new line

File: portfolio/src/main/webapp/projects/commentart/scripts/replyform.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  getFormContent() {
+    return '<label class="form-label" for="responder-name">Name</label>' +
+        '<input type="text" name="responder-name"/><label class="form-label" for="response">Reply</label>' +
+        '<input type="text" name="response"/><input type="submit" value="Submit"/>';
+  }
+
+  enableAsyncReply(formElement, addReply) {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/replyform.js]: Async is an implementation detail and the name of a function should only be describing what it does, instead of how it does it. The caller shouldn't need to know whether or not the implementation is async or not.

File: portfolio/src/main/webapp/projects/commentart/scripts/repliesmanager.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+class RepliesManager {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/repliesmanager.js:1]: Missing doc for the class.

File: portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+class CommentsManager {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js:1]: Missing doc for class.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
+class Comment {
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/comment.js:1]: Missing doc for class

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    this.repliesManager = repliesManager;
+  }
+
+  /** Converts a comment that is in .json format to a `Comment` object. */
+  static toComment(commentJson) {
+    const commenterName = commentJson.name;
+    const commentMsg = commentJson.commentMsg;
+    const commentSentiments = commentJson.sentiment;
+    const repliesManager = RepliesManager.toRepliesManager(commentJson.repliesManager);
+    return new Comment(commenterName, commentMsg, commentSentiments, repliesManager);
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: The name `repliesManager` is a bit confusing, how can some json data be able to manage something? AFAICS, repliesManager in the json is an id, and a list of `Reply`s.

File: portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  constructor(comments) {
+    this.comments = comments;
+    this.commentsElement = document.getElementById('comments-placeholder');
+    // TODO: ensure that commentsElement exists
+  }
+
+  addComment(replyJson) {
+    const newComment = Comment.toComment(replyJson);
+    this.comments.push(newComment);
+    this.commentsElement.appendChild(newComment.toElement);
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js:12]: What's the purpose for CommentsManager to hold a reference of commentsElement?

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  body.onload = fetchAndShowComments;
+}
+
+/** Fetches comments from the server. */
+const fetchAndShowComments = () => {
+  fetch('/read_comments')
+      .then(response => response.json())
+      .then(commentsJson => {
+        const comments = commentsJson.map(Comment.toComment(commentJson));
+        const commentsManager = new CommentsManager(comments);
```
Comment (fanwgwg) [inline portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Why do we need to create a new CommentsManager for every fetch request?

**Review by fanwgwg — CHANGES_REQUESTED** [review]
I can see a great effort towards encapsulation and abstraction, good work! 

But generally I think View / Data Model / Logic are kind of mixed together, and I would suggest a better separation between them, and also think about how data flows between each other, and how many instances of each class/manager should actually coexist.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 1 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:8]: Ahh.. Okay. But I don't intend to allow for more values. There is a reason for this. Thanks for pointing it out though!

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 3 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:41]: Oops. I called the wrong function. Thanks for pointing it out! (:

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: This fetch request is done to load data from the server. It will only be called once the the page reloads. I suppose, I need a manager for each session, don't I? :O

File: portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js (hunk 11 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/commentsmanager.js:12]: Yup! And to summarise the result. I have not added that interface/class for that though.

**Comment by pohlinwei** [comment]
I pushed the code that is meant for a separate PR by accident. I am so sorry. 😥

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 2 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:25]: Hmm... Is it? I am only creating a `li` element and returning it. Did I miss out something?

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Sorry about the confusion. My intention is to have `repliesManager` manage the `Reply`s. I need its `id` because I want to set `Reply`'s ancestor to `RepliesManager`. Since this is confusing, how do you think I can better implement it? Thanks!

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 1 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:8]: Oh.. Wait. I realised that I read the document incorrectly. Will change this to allow a range of values. Thanks for pointing out! (:

File: portfolio/src/main/java/com/google/sps/servlets/ReadCommentsServlet.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    List<Comment> comments = new ArrayList<>();
+    for (Entity commentEntity : results.asIterable()) {
+      comments.add(toComment(commentEntity));
+    }
+    String json = gson.toJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  private Comment toComment(Entity commentEntity) {
```
Comment (fanwgwg) [inline portfolio/src/main/java/com/google/sps/servlets/ReadCommentsServlet.java:44]: These `to` methods should not be in the scope of ReadCommentsServlet, instead, they should be the knowledge of each data class.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 1 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:8]: I don't see changes being made for this, please leave the comment unresolved until change is made :)

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: If I'm understanding correctly, each comment has a corresponding RepliesManager, identified by id. I'm wondering that instead of keeping the id in replied manager, can we have an id for each comment instead? The comment json could simply contain a list of repliles, and as of now I don't see the importance of RepliesManager, as it's only handling the logic of adding comment, which I feel that could be inside `Comment` class.

For example, Comment class will have the following
```
public class Comment {
  private String id;
  private String commenterName;
  private String commentMsg;
  private float sentiment;
  private List<String> replies;

  public Comment(String id, String commenterName, String commentMsg, float sentiment, List<String> replies) {
    this.id = id;
    this.commenterName = commenterName;
    this.commentMsg = commentMsg;
    this.sentiment = sentiment;
    this.replies = replies;
  }

  // Constructs comment fro entity
  public Comment(Entity commentEntity) {
     ...
  }

  public void addReply(...) {
    ...
  }
 }
```

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 2 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:25]: Oh I see, that makes sense, thanks for clarifying

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: And then on the front-end side, can refer to my comment below, as I was suggesting removing the corresponding classes for front-end side

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: The name `CommentsManager` is a bit confusing, it seems to be managing something but actually it's only responsible for creating some UI elements. I think that on the front-end side, we don't need so many abstraction for UIs, because they're just UI logic. Front-end should only get data from server, and display accordingly. 

In terms of creating a separate class for Comment, Reply, I don't find them currently being necessary for front-end. Each time we create a separate class, it must be that we need different instantiations of it and we need to keep the reference of it for future use, which I don't find to be super convincing right now. Otherwise this will only increase more memory usage by keeping UI elements in the memory.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 1 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:8]: Oh I saw it, sorry I was misreading the comments. Resolving this

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Thanks for the suggestion! (: This was what I had in mind. However, after brooding over it, I realised that I don't necessary want `Comment` to know about the replies. I don't think that it needs to know about it. Instead, for this application, `Comment` simply needs to know the number of positive/negative replies. I would like to use the sentiments from these replies and the comment's sentiment as data for this app. The number of replies with positive/negative sentiment will be managed by `RepliesManager`.  

What do you think? 😅

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Hmm... I see where you are coming from. Perhaps the above comment (which I just posted) is relevant here. 

Oh okay! But if I do not maintain classes, how do I ensure that when the user clicks a reply button, it will be added to the correct comment? Do it using numbers? In addition, how do I keep track of the number of positive/negative replies for each comment?

File: portfolio/src/main/java/com/google/sps/servlets/ReadCommentsServlet.java (hunk 13 above)
Comment (pohlinwei) [reply portfolio/src/main/java/com/google/sps/servlets/ReadCommentsServlet.java:44]: Okay. Thanks!

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Hmm I don't quite see what you mean.

The current logic is that Comment holds a RepliesManager, which is capable of addReply. If someone wants to addReply to the RepliesManager, we would do `comment.getRepliesManger().addReply()`. From the outside point of view, the knowledge of replies is still scoped to comment, and so far I don't quite see the need for ReplyManager, besides the fact that it doesn't actually have any logic for *managing*. It's only holds some replies data. wdyt?

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: I made the above suggestion based on the fact that your comment doesn't have an id. Generally what I would prefer as a better solution on a high level is that, your Comment class contains a String id, with all the relevant data for the comment, including replies (It's okay for a reply itself to be a class, as it's some self-contained data), and this will be the single source of truth.

On the front end side, you could have a method which creates a comment UI element from the Comment data, and its id will be used in the onClick listener. Each time the onClick listener triggered, you'll be talking to the server by providing the id, which can be used to identify the comment, as well as replies, because each reply can only belong to one comment, which is identified by the id.

For number of positive/negative replies, similarly, it is kept inside the Comment data. Your front end basically should not have much logic. It should only display according to data, and give server some updates through the unique id, and then update again when new data fetched.

Does this make sense?

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Sounds good! But this means that `Comment` will own the `ReplyForm` as well, won't it? Presently, `ReplyForm` is coupled to `RepliesManager`.

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: I see... It does! But I would like to constantly change the display (CSS shape) based on the sentiments. Does this mean that I should constantly perform queries? Personally, I feel that it will be a waste to perform the queries. Hence, I decided to weave the data component into the front-end. Thanks in advance! :D

**Comment by pohlinwei** [comment]
Shall I close this PR and do a new one? I feel that the changes will be quite major. And there is no point developing on this PR. 😅 I appreciate the comments though! I will definitely improve on what you have mentioned here, if they remain relevant. (:

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Hmmm how would you know about sentiment if you don't send it back to server? I guess you want to perform the sentiment analysis on front end?

What I would suggest is that the sentiment should be peroformed by the server, otherwise you may need to expose your api key on the front end which can be easily hacked in by others. So the flow would be when user replies, you would send the reply back to server, and wait for server to return you the sentiment of the reply, and then change the UI based on the result. Does this make sense?

**Comment by fanwgwg** [comment]
Sure, feel free to close this PR and open a new one if that's cleaner for you! I hope this doesn't block your demo tmr :( Feels like I might be blocking you too much

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Hmmm I was talking about the server side setup here... See my example code above, it's in java :) But I should have commented under the java code instead.

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Oh no. I do want it to be done on the server side. However, I want the results to be saved on the front end. Specifically, for each comment, I would like to know that it has X number of positive replies and Y number of negative replies. In addition, I want to know whether the comment is positive/negative. 

In fact, I don't really need to store the replies. However, I did so in this PR because I don't want to complicate matters before I know that it works. 😅 

Nonetheless, thanks a lot for the advice! :D

What are your thoughts about the following?
- each comment will store its own sentiment, its id and the number of negative/positive replies
- a sentiment manager will use the sentiment data from the comments (I have been wanting to implement this manager but I wanted to make sure things work before going ahead with it)

I am not sure whether this will be how it pans out but I will try. 😂

**Comment by pohlinwei** [comment]
Oh no. It is fine. (: I am largely to blame: I didn't manage my time well. Besides, I genuinely appreciate your comments! :D 

Frankly, this is all experimental for me. I haven't really used OOP for front-end previously. However, I found it difficult to keep track of what is going on, especially if the bulk of the code remains as functions. 😂 Therefore, I was wondering if this would help.

File: portfolio/src/main/webapp/projects/commentart/scripts/comment.js (hunk 10 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/comment.js:21]: Oh. HAHA. Sorry. I didn't think too much about the language. I just took it as pseudocode. 😅

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Yes the Comment looks good to me. I do see why things are getting a bit messy right now, but that is indeed a problem with vanila JS, so no worries if you find this not as trivial or elegant. If you later started to grab some web frametwork, like ReactJS or AngularJS, you'll find things much easier than this :)

Can you be more specific on what SentimentManager does?

**Comment by fanwgwg** [comment]
Yup as I mentioned, this is all about your learning, so feel free to explore things around and try out different approaches to one problem!

File: portfolio/src/main/webapp/projects/commentart/scripts/main.js (hunk 12 above)
Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: I see! HAHA. Yes. I have been wanting to learn a web framework (I did it very long ago but I didn't really appreciate then because I was building extremely basic websites). :D 

I have thought through the architecture in greater detail and decided that I will use the numerical values of the sentiments instead of positive/negative. Here is the architecture:

- `ColourUtilities` converts a sentiment value to a colour (via basic addition/multiplication)
- `SentimentColour` is an interface with `getColour` method
- `MultiSentimentColours` extends `Sentiment` and has `getResponseColour`(which returns an array of `SentimentColour`s) and `getId` methods
- `Reply` implements `SentimentColour` and has message, name and sentimentColour properties
- `Comment` implements `MultiSentiments` and has id, replies (an array of `Reply`s), sentimentColour, message and name properties 
- `Painter` takes in an array of `MultiSentiments` and computes the desired colour for each `MultiSentiment`

Basically, I would like to have the following data for my front-end (besides the name/message): a colour for each comment, a colour for each reply, and a colour for each comment + reply (i.e. the overall response). The latter colours are provided by `Painter`.

Comment (fanwgwg) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: A few things to notice:

1. How do you plan to have interface defined in javascript?
2. I think there's some misuse/misunderstanding about interface here. Interface should not be used when you found that two classes will have same methods, it should be used when you want different implementations of one object. For example, why would Reply implement SentimentColour? Reply itself is not some kind of color. Similarly, Comment is not a sentiment. The same things applies to MultiSentimentColours and Sentiment, how would a color extend a sentiment?

The main point is that we shouldn't over architect and force something that aren't relevant to be grouped together. Reply can have a color, Comment can have some colors, but they don't necessarily to be implementing an interface to have such a getColor/getResponseColour method, as we don't see a point where interface is important and concrete implementation is not. 

Think about List and ArrayList, the reason why we have List is that we don't want people to care about the implementations of the List, all they need to know is that a List can have these and these functionalities. Take Reply for example, we don't just care about the SentimentColour of it, we also care about its message, its name etc, same for Comment.

Based on this, do you have some neater design on it :)

Comment (pohlinwei) [reply portfolio/src/main/webapp/projects/commentart/scripts/main.js:13]: Thanks! Sorry, I just saw this.><  I am doing away with the interfaces. While implementing it, I realised that they are not necessary. I guess I was a tad too stuck with `Comment` and `Reply` when in fact I don't really need them -- besides for the conversion of JSON string.

Instead, I think what is more relevant is to create a `Palette` class that uses the data from each reply and comment. `Painter` will then store the `Palette`s. My reply and comment classes will just be used to create the formatted JSON string. 

On the front-end, I removed all the classes and I am only using function calls. 😅
